src/System.IO.Pipelines/src/System/IO/Pipelines/PipeOptions.cs

```diff
@@ -0,0 +1,79 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+
+namespace System.IO.Pipelines
+{
+    /// <summary>
+    /// Represents a set of <see cref="Pipe"/> options
+    /// </summary>
+    public class PipeOptions
+    {
+        /// <summary>
+        /// Default instance of <see cref="PipeOptions"/>
+        /// </summary>
+        public static PipeOptions Default { get; } = new PipeOptions();
+
+        /// <summary>
+        /// Creates a new instance of <see cref="PipeOptions"/>
+        /// </summary>
+        public PipeOptions(
+            MemoryPool<byte> pool = null,
+            PipeScheduler readerScheduler = null,
+            PipeScheduler writerScheduler = null,
+            long pauseWriterThreshold = 0,
+            long resumeWriterThreshold = 0,
+            int minimumSegmentSize = 2048)
+        {
+
+            if (resumeWriterThreshold < 0 && resumeWriterThreshold > pauseWriterThreshold)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.resumeWriterThreshold);
+            }
+
+            if (pauseWriterThreshold < 0)
```

Comment:
You'll have a better chance of setting the ParamName correctly if you perform this validation check before the `resumeWriterThreshold` check.